libcloud/compute/drivers/kubevirt.py

```diff
@@ -391,63 +772,162 @@ def create_node(
+                warnings.warn(
+                    "The disk type {} is not tested. Use at your own risk.".format(disk_type)
+                )
+                volumes_dict = {disk_type: disk.get("volume_spec", {}), "name": disk_name}
+
             disk_dict = {device: {"bus": bus}, "name": disk_name}
             vm["spec"]["template"]["spec"]["domain"]["devices"]["disks"].append(disk_dict)
             vm["spec"]["template"]["spec"]["volumes"].append(volumes_dict)
+        # end of for disk in ex_disks
 
+        # image -> containerDisk
         # adding image in a container Disk
         if isinstance(image, NodeImage):
             image = image.name
 
-        volumes_dict = {"containerDisk": {"image": image}, "name": "boot-disk"}
-        disk_dict = {"disk": {"bus": "virtio"}, "name": "boot-disk"}
-        vm["spec"]["template"]["spec"]["domain"]["devices"]["disks"].append(disk_dict)
-        vm["spec"]["template"]["spec"]["volumes"].append(volumes_dict)
+        boot_disk_name = "boot-disk-" + str(uuid.uuid4())
+        volumes_dict = {"containerDisk": {"image": image}, "name": boot_disk_name}
+        disk_dict = {"disk": {"bus": "virtio"}, "name": boot_disk_name}
+        # boot disk should be the first one, otherwise it will not boot
+        vm["spec"]["template"]["spec"]["domain"]["devices"]["disks"].insert(0, disk_dict)
+        vm["spec"]["template"]["spec"]["volumes"].insert(0, volumes_dict)
+
+        # auth -> cloud-init
+        if auth is not None:
+            # auth requires cloud-init,
+            # and only one cloud-init volume is supported by kubevirt.
+            # So if both auth and cloud-init are provided, raise an error.
+
+            for volume in vm["spec"]["template"]["spec"]["volumes"]:
+                if "cloudInitNoCloud" in volume or "cloudInitConfigDrive" in volume:
+                    raise ValueError(
+                        "Setting auth and cloudInit at the same time is not supported."
+                        "Use deploy_node() instead."
+                    )
+
+            # cloud-init volume
+            cloud_init_volume = "auth-cloudinit-" + str(uuid.uuid4())
+            disk_dict = {"disk": {"bus": "virtio"}, "name": cloud_init_volume}
+            volume_dict = {
+                "name": cloud_init_volume,
+                "cloudInitNoCloud": {"userData": ""},
+            }
+
+            # auth
+            # cloud_init_config reference: https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#injecting-ssh-keys-with-cloud-inits-cloud-config
+            if isinstance(auth, NodeAuthSSHKey):
+                public_key = auth.pubkey
+                cloud_init_config = (
+                    """#cloud-config\n""" """ssh_authorized_keys:\n""" """  - {}\n"""
+                ).format(public_key)
+            elif isinstance(auth, NodeAuthPassword):
+                password = auth.password
+                cloud_init_config = (
+                    """#cloud-config\n"""
```

Comment:
It would be a bit more readable if we stored cloud init configs in template files and then load those + render them here.